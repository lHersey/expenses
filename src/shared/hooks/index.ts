export * from './useColorAdjust';
export * from './useTabScreenOptions';
