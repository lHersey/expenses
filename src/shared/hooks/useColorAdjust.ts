import { useMemo } from 'react';

const useColorAdjust = (color: string, modifier: number) => {
  return useMemo(
    () =>
      '#' +
      color
        .replace(/^#/, '')
        .replace(/../g, c => ('0' + Math.min(255, Math.max(0, parseInt(c, 16) + modifier)).toString(16)).substr(-2)),
    [color, modifier],
  );
};

export const colorAdjust = (color: string, modifier: number) =>
  '#' +
  color
    .replace(/^#/, '')
    .replace(/../g, c => ('0' + Math.min(255, Math.max(0, parseInt(c, 16) + modifier)).toString(16)).substr(-2));

export { useColorAdjust };
