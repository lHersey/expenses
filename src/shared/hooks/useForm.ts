import { FormikErrors, FormikHelpers, FormikTouched, useFormik } from 'formik';
import { createRef, RefObject, useCallback, useMemo } from 'react';
import * as yup from 'yup';

type FormType<T extends Record<string, unknown>> = T;

type Focusable = { focus: () => void };

type FormOptions<T extends Record<string, unknown>> = {
  initialValues: T;
  validationSchema: yup.AnyObjectSchema;
  validateOnMount?: boolean;
  clearAfterSubmit?: boolean;
  onSubmit: (values: FormType<T>, formikBag: FormikHelpers<T>) => void;
};

export const useForm = <T extends Record<string, unknown>>(options: FormOptions<T>) => {
  const { initialValues, onSubmit, validationSchema, validateOnMount, clearAfterSubmit } = options;

  const refs = useRefsMemo<T>(initialValues);

  const handleSubmitEditing = useCallback(
    (fieldToFocus?: keyof T) => () => {
      if (!fieldToFocus) return;

      refs[fieldToFocus]?.current?.focus();
    },
    [refs],
  );

  const handleRef = useCallback((name: keyof T) => refs[name], [refs]);

  const form = useFormik({
    initialValues,
    onSubmit,
    validationSchema,
    validateOnMount,
  });

  const handleSubmit = useCallback(async () => {
    await form.submitForm();
    if (clearAfterSubmit) {
      form.resetForm(initialValues);
    }
  }, [form, clearAfterSubmit, initialValues]);

  const errors = useTouchedErrors<T>(form.errors, form.touched);

  return {
    errors,
    refs,
    values: form.values,
    isValid: form.isValid,
    isSubmitting: form.isSubmitting,
    resetForm: form.resetForm,
    handleBlur: form.handleBlur,
    handleChange: form.handleChange,
    setFieldValue: form.setFieldValue,
    handleSubmit,
    handleSubmitEditing,
    handleRef,
  };
};

//We will pass only errors for touched inputs
function useTouchedErrors<T extends Record<string, unknown>>(errors: FormikErrors<T>, touched: FormikTouched<T>) {
  return Object.keys(errors).reduce((memo, key) => {
    memo[key as keyof T] = touched[key] ? (errors[key] as string) : undefined;

    return memo;
  }, {} as Record<keyof T, string | undefined>);
}

// We will create a ref for each object in the form (Each key of initial values)
function useRefsMemo<T extends Record<string, unknown>>(initialValues: T) {
  return useMemo(() => {
    return Object.keys(initialValues).reduce((memo, key) => {
      memo[key as keyof T] = createRef<Focusable>();
      return memo;
    }, {} as Record<keyof T, RefObject<Focusable>>);
  }, [initialValues]);
}
