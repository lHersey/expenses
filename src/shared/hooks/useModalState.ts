import { useCallback, useState } from 'react';

const useModalState = (): [boolean, () => void] => {
  const [isVisible, setVisible] = useState(false);
  const handleShowModal = useCallback(() => setVisible(v => !v), []);

  return [isVisible, handleShowModal];
};

export { useModalState };
