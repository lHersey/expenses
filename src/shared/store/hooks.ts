import { Action, ThunkAction, ThunkDispatch } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector as useReduxSelector } from 'react-redux';

import rootReducer from './reducers';

export type RootState = ReturnType<typeof rootReducer>;
export type AppThunk<ReturnType = void> = ThunkAction<Promise<ReturnType>, RootState, null, Action<string>>;
export type ReduxDispatch = ThunkDispatch<RootState, null, Action<string>>;

export const useThunkDispatch = () => useDispatch<ReduxDispatch>();
export const useSelector: TypedUseSelectorHook<RootState> = useReduxSelector;
