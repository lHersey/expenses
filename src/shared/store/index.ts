import AsyncStorage from '@react-native-async-storage/async-storage';
import { configureStore } from '@reduxjs/toolkit';
import { persistReducer, persistStore } from 'redux-persist';

import rootReducer from './reducers';

const persistedReducer = persistReducer(
  {
    key: 'PARKING',
    storage: AsyncStorage,
  },
  rootReducer,
);

const store = configureStore({
  reducer: persistedReducer,
  devTools: __DEV__,
  middleware: getDefaultMiddleware => getDefaultMiddleware({ serializableCheck: false }),
});

const persistor = persistStore(store);

export { persistor };
export default store;
