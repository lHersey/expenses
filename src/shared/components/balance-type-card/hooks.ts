import { useMemo } from 'react';
import { BalanceType } from 'shared/enum';
import { IoniconsName } from 'shared/types/icons';
import { useTheme } from 'styled-components/native';

type BalanceTypeData = {
  label: string;
  icon: IoniconsName;
  color: string;
};

export const useBalanceTypeInformation = (type: BalanceType) => {
  const theme = useTheme();
  const balancePayload = useMemo(
    () =>
      new Map<BalanceType, BalanceTypeData>([
        [BalanceType.EXPENSE, { label: 'Expenses', icon: 'arrow-up-circle-outline', color: theme.ERROR_COLOR }],
        [BalanceType.INCOME, { label: 'Income', icon: 'arrow-down-circle-outline', color: theme.SUCCESS_COLOR }],
      ]),
    [theme],
  );

  return balancePayload.get(type);
};
