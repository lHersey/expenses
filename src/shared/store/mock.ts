import { EntityState } from '@reduxjs/toolkit';
import { BalanceType } from 'shared/enum';
import Category from 'shared/models/category';

export const mockCategory = (category: Partial<Category> = {}): Category => {
  return {
    ...new Category('1', 'CategoryName', '#d3fae2', BalanceType.INCOME, 'add-outline'),
    ...category,
  };
};

export const mockCategoryEntity = (...categories: Category[]) => {
  return categories.reduce(
    (entityState, category) => {
      entityState.entities[category.id] = category;
      entityState.ids.push(category.id);

      return entityState;
    },
    { entities: {}, ids: [] } as EntityState<Category>,
  );
};
