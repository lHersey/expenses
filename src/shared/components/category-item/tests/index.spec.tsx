import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import Category from 'shared/models/category';
import Providers from 'shared/providers/test-providers';
import { RootState } from 'shared/store/hooks';
import { mockCategory, mockCategoryEntity } from 'shared/store/mock';

import CategoryItem, { CATEGORY_ITEM_TEST_ID } from '../index';

describe('<CategoryItem />', () => {
  let state: Partial<RootState>;
  let category: Category;

  beforeEach(() => {
    category = mockCategory();
    state = {
      category: mockCategoryEntity(category),
    };
  });

  describe('snapshots', () => {
    it('should match snapshot', () => {
      expect.hasAssertions();
      const rendered = render(
        <Providers state={state}>
          <CategoryItem categoryId={category.id} onPress={jest.fn()} />
        </Providers>,
      ).toJSON();

      expect(rendered).toMatchSnapshot();
    });
  });

  describe('unit', () => {
    it('should render the text', () => {
      expect.hasAssertions();

      const { queryByText } = render(
        <Providers state={state}>
          <CategoryItem categoryId={category.id} onPress={jest.fn()} />
        </Providers>,
      );

      expect(queryByText(category.name)).toBeTruthy();
    });

    it('should handle onPress', () => {
      expect.hasAssertions();

      const handleOnPressMock = jest.fn();

      const { getByTestId } = render(
        <Providers state={state}>
          <CategoryItem categoryId={category.id} onPress={handleOnPressMock} />
        </Providers>,
      );

      const pressable = getByTestId(CATEGORY_ITEM_TEST_ID);
      fireEvent.press(pressable);

      expect(handleOnPressMock).toHaveBeenCalledWith(category.id);
    });
  });
});
