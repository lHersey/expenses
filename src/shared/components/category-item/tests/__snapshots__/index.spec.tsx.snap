// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryItem /> snapshots should match snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#f2f5f7",
          "borderRadius": 20,
          "flexDirection": "row",
          "flexGrow": 1,
          "marginTop": 10,
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
        },
      ]
    }
    testID="CATEGORY_ITEM"
  >
    <View
      color="#d3fae2"
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#f1ffff",
            "borderRadius": 25,
            "height": 50,
            "justifyContent": "center",
            "marginRight": 20,
            "width": 50,
          },
        ]
      }
    >
      <Text />
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#5e5e5e",
            "fontFamily": "Inter_500",
            "fontSize": 16,
          },
          Object {
            "fontFamily": "Inter_300",
          },
        ]
      }
    >
      CategoryName
    </Text>
  </View>
</RNCSafeAreaProvider>
`;
